src/H5Oalloc.c

```diff
@@ -2269,6 +2262,16 @@ H5O__condense_header(H5F_t *f, H5O_t *oh)
     /* check args */
     assert(oh != NULL);
 
+    /* First remove all deleted messages from the object header */
+    for (unsigned u = 0; oh->num_deleted_mesgs > 0 && u < oh->nmesgs;)
+        if (oh->mesg[u].type->id == H5O_DELETED_ID) {
+            memmove(&oh->mesg[u], &oh->mesg[u + 1], ((oh->nmesgs - 1) - u) * sizeof(H5O_mesg_t));
```

Comment:
I notice the old logic was effectively checking `u < (oh->nmesgs - 1)`. What if `u == (oh->nmesgs - 1)` here?

Reply:
In that case it will calculate the address of the first element past the used size of the array array but the size will calculate to zero so it will not read any bytes past the end of the used bytes in the array. We could add a check but I don't think it's necessary